Support protobuf format in QP's metrics exchange

skonto opened the pull request:
Fixes #8378

<!-- Please include the 'why' behind your changes if no issue exists -->
## Proposed Changes

* It is based on previous work from here #8396.
* Supports by default the protobuf based format for the QP <-> Autoscaler communication.
* Introduces a proto stat reporter for the QP. Note here that it does not use the Prometheus client model and its related exposition format based on Proto 2, as there is no support for [proto 3](https://github.com/prometheus/docs/issues/549) coming and in general Prometheus and its client libs are [moving away](https://github.com/prometheus/docs/issues/549#issuecomment-559560195) from [protobuf](https://github.com/prometheus/client_model).

* Keeps the current Prometheus text format for migration scenarios.
* It does not do any compression of the proto metrics struct since the are no real benefits unless there is an unlike case where `podName` in `Stat` is highly compressible and contains a lot of repetitions. 

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```
NONE
```

/assign @markusthoemmes

skonto commented on pkg/autoscaler/metrics/http_scrape_client_test.go:
```diff
@@ -186,9 +219,54 @@ func TestHTTPScrapeClient_Scrape_ErrorCases(t *testing.T) {
+			} {
+				// only the first two cases apply for the proto based response, unlike Prometheus gathered statistics
+				// proto will not have any missing values.
+				if i >= 2 && j == 1 {
```
For proto metrics there is no scenario where metrics does not exist. The is only the case for the 
Prometheus client library where it does a [best effort](https://github.com/prometheus/client_golang/blob/b0cdec211b8bf9d5b1edabff4d32088a11f18310/prometheus/registry.go#L150) in case of an error.

markusthoemmes replied:
Hmm, let's add an explicit `bool` or something to indicate these shouldN't run on protobuf code or something. This feels super cryptic.

Or just copy & paste only the relevant cases to the test below?

markusthoemmes commented on pkg/autoscaler/metrics/http_scrape_client.go:
```diff
@@ -45,19 +48,42 @@ func (c *httpScrapeClient) Scrape(url string) (Stat, error) {
+
+	// Ask for protobuf by default. Note that during migration this will not trigger the proto format supported
+	// by Prometheus reporter as the latter uses `application/vnd.google.protobuf`.
+	req.Header.Add("Accept", "application/protobuf")
```
Does this keep the auto-gzip handling for the prometheus handler intact or do we need to "manually" do that now?

skonto replied:
@markusthoemmes afaik the transport will ask for `gzip` by [itself.](https://golang.org/pkg/net/http/#Transport) when there is no accept-encoding header present. The Prometheus handler properly handles this in this [code part.](https://github.com/prometheus/client_golang/blob/b0cdec211b8bf9d5b1edabff4d32088a11f18310/prometheus/promhttp/http.go#L157) which checks if there is a [gzip encoding](https://github.com/prometheus/client_golang/blob/b0cdec211b8bf9d5b1edabff4d32088a11f18310/prometheus/promhttp/http.go#L356) in the hearder section. In the proto case gzip will not be triggered so that means we need to explicitly add gzip on in the request header before it is received by that handler.

markusthoemmes commented on pkg/autoscaler/metrics/http_scrape_client.go:
```diff
@@ -45,19 +48,42 @@ func (c *httpScrapeClient) Scrape(url string) (Stat, error) {
-	return extractData(resp.Body)
+	if resp.Header.Get("Content-Type") == "application/protobuf" {
+		return extractDataFromProto(resp.Body)
+	} else {
```
Golang nit: Drop the else branch and just return.

markusthoemmes commented on pkg/autoscaler/metrics/http_scrape_client.go:
```diff
@@ -45,19 +48,42 @@ func (c *httpScrapeClient) Scrape(url string) (Stat, error) {
+}
+
+func extractDataFromProto(body io.Reader) (Stat, error) {
+	stat := emptyStat
```
I'd just use `stat := &Stat{}` here and then call its `Unmarshal` function directly below? :thinking:

markusthoemmes commented on pkg/queue/prometheus_stats_reporter_test.go:
```diff
@@ -34,73 +34,21 @@ const (
-		expectedConcurrency         float64
-		expectedProxiedConcurrency  float64
-	}{{
+func GetReporterReportTests() []ReporterTestInput {
```
No need to wrap them in a function I think. You could also just do:

```
var testCases := []struct { ... }
```

and then use that list in both functions.

skonto replied:
@markusthoemmes I thought both of the options you listed above, but it depends on code practices, I added the comment to clarify what it should run. Anyway I will copy the test cases as it is the easier although more verbose option.

markusthoemmes commented on pkg/queue/protobuf_stats_reporter.go:
```diff
@@ -0,0 +1,104 @@
+}
+
+func httpError(rsp http.ResponseWriter, err error) {
+	rsp.Header().Del(contentEncodingHeader)
```
Why is this necessary?

skonto replied:
@markusthoemmes this follows this [part](https://github.com/prometheus/client_golang/blob/4739e9cf8cd673602da0facb331c6b8393559b46/prometheus/promhttp/http.go#L373). I wanted to make sure that the client knows that plain text is returned, removing any headers previously set.

markusthoemmes commented on pkg/queue/protobuf_stats_reporter.go:
```diff
@@ -0,0 +1,104 @@
+		stat := r.stat.Load()
+		if stat == nil {
+			httpError(rsp, errors.New("no metrics available yet."))
+		} else {
```
Golang nit: Drop the else branch and return early from the error branch above.

markusthoemmes commented on pkg/queue/protobuf_stats_reporter.go:
```diff
@@ -0,0 +1,104 @@
+		if stat == nil {
+			httpError(rsp, errors.New("no metrics available yet."))
+		} else {
+			w := io.Writer(rsp)
```
Is this cast necessary?

skonto replied:
Nope I will remove it.

markusthoemmes commented on pkg/queue/protobuf_stats_reporter.go:
```diff
@@ -0,0 +1,104 @@
+			header := rsp.Header()
+			header.Set(contentTypeHeader, "application/protobuf")
+			data := stat.(metrics.Stat)
+			buffer, err := proto.Marshal(&data)
+			if err != nil {
+				httpError(rsp, err)
+			}
```
Let's move this before setting the header to make unsetting it unnecessary.

markusthoemmes commented on pkg/queue/protobuf_stats_reporter.go:
```diff
@@ -0,0 +1,104 @@
+			if err != nil {
+				httpError(rsp, err)
+			}
+			_, err = w.Write(buffer)
+			if err != nil {
+				httpError(rsp, err)
+			}
```
I believe golang servers generally don't handle the write error here as this will always mean that the connections broken anyway AFAIK. The http error is likely never received. I'd drop the error handling.

skonto replied:
I had a second thought on this yes most of the time it may not even be feasible to talk to client if things fail at this point but usually logging is something that could be proven useful.

markusthoemmes replied:
Sure, if it's not already logged somewhere anyway.

markusthoemmes commented on pkg/queue/protobuf_stats_reporter_test.go:
```diff
@@ -0,0 +1,86 @@
+func checkProtoData(t *testing.T, got float64, want float64) {
+	t.Helper()
+	if got != want {
+		t.Errorf("Got %v for Gauge value, wanted %v", got, want)
```
You're not talking gauge values here. I'd just do `Got = %v, want %v` instead.

markusthoemmes commented on cmd/queue/main.go:
```diff
@@ -232,13 +233,20 @@ func main() {
 	}
 
+	protoStatReporter, err := queue.NewProtobufStatsReporter(env.ServingPod, reportingPeriod)
+
```
nit: drop the empty line here.

markusthoemmes commented on pkg/autoscaler/metrics/http_scrape_client.go:
```diff
@@ -45,19 +48,42 @@ func (c *httpScrapeClient) Scrape(url string) (Stat, error) {
 	}
 
-	return extractData(resp.Body)
+	if resp.Header.Get("Content-Type") == "application/protobuf" {
```
Put `"application/protobuf"` in a shared constant?

skonto replied:
I thought of that but I didn't find a nice place for constants. I will fix it.

markusthoemmes commented:
As a general comment: This PR is fairly large. Does it make sense to cut it into 2-3 PRs maybe?

I could imagine the following order:
1. Add the protobuf server (including tests) [don't hook it up yet]
2. Add the protobuf parsing code (including tests), but don't set the Accept header yet.
3. Hook up the protobuf server and start setting the Accept header.

markusthoemmes commented:
/ok-to-test

skonto commented on pkg/autoscaler/metrics/http_scrape_client.go:
```diff
@@ -45,19 +48,42 @@ func (c *httpScrapeClient) Scrape(url string) (Stat, error) {
+	}
+}
+
+func extractDataFromProto(body io.Reader) (Stat, error) {
```
sure.

skonto commented on pkg/queue/prometheus_stats_reporter_test.go:
```diff
@@ -151,7 +99,62 @@ func TestReporterReport(t *testing.T) {
+		revision  string
+		pod       string
+	}{{
+		"Empty_Namespace_Value",
```
@vagababov this is code that pre-existed, I can try improve it.

taragu commented on pkg/autoscaler/metrics/http_scrape_client_test.go:
```diff
@@ -82,64 +107,72 @@ func TestNewHTTPScrapeClient_ErrorCases(t *testing.T) {
+		if err != nil {
+			t.Errorf("scrapeViaURL = %v, want no error", err)
+		}
+		if stat.AverageConcurrentRequests != 3.0 {
```
the numbers in the rest of this test (and TestHTTPScrapeClientScrapeHappyCaseWithOptionals) can be replaced with the defined constants

vagababov commented on pkg/autoscaler/metrics/http_scrape_client.go:
```diff
@@ -20,12 +20,17 @@ import (
 )
 
+const (
+	protoAcceptContent = "application/protobuf"
```
can't we use the same constant?
We can put it in `pkg/network` or thereabouts.

skonto replied:
Ok makes sense I didnt want to put it in an arbitrary pkg.

vagababov commented on pkg/autoscaler/metrics/http_scrape_client.go:
```diff
@@ -45,19 +50,42 @@ func (c *httpScrapeClient) Scrape(url string) (Stat, error) {
+	}
+	err = stat.Unmarshal(bodyBytes)
+	if err != nil {
+		return emptyStat, fmt.Errorf("Unmarshalling failed: %w", err)
```
```suggestion
		return emptyStat, fmt.Errorf("unmarshalling failed: %w", err)
```

vagababov commented on pkg/autoscaler/metrics/http_scrape_client_test.go:
```diff
@@ -204,13 +237,61 @@ func TestHTTPScrapeClient_Scrape_ErrorCases(t *testing.T) {
 	}
 }
 
+func TestHTTPScrapeClient_ScrapeProto_ErrorCases(t *testing.T) {
```
```suggestion
func TestHTTPScrapeClientScrapeProtoErrorCases(t *testing.T) {
```
and elsewhere

vagababov commented on pkg/autoscaler/metrics/http_scrape_client_test.go:
```diff
@@ -204,13 +237,61 @@ func TestHTTPScrapeClient_Scrape_ErrorCases(t *testing.T) {
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			hClient := newTestHTTPClient(getProtoHTTPResponse(test.responseCode, test.responseContext, protoAcceptContent), test.responseErr)
+			sClient, err := newHTTPScrapeClient(hClient)
+			if err != nil {
+				t.Errorf("newHTTPScrapeClient=%v, want no error", err)
+			}
+			if _, err := sClient.Scrape(testURL); err != nil {
+				if !strings.Contains(err.Error(), test.expectedErr) {
+					t.Errorf("Got error message: %q, want to contain: %q", err.Error(), test.expectedErr)
+				}
+			} else {
+				t.Error("Expected error from newServiceScraperWithClient, got nil")
+			}
+		})
+	}
```
```suggestion
			sClient, err := newHTTPScrapeClient(hClient)
			if err != nil {
				t.Fatalf("newHTTPScrapeClient=%v, want no error", err)
			}
			_, err := sClient.Scrape(testURL)
			if err == nil {
				t.Fatal("Got no error")
			}
			if !strings.Contains(err.Error(), test.expectedErr) {
				t.Errorf("Error = %q, want to contain: %q", err.Error(), test.expectedErr)
			}
		})
	}
```

skonto replied:
@vagababov it seems that this code part pre-existed, should we do these kind of refactorings in this PR?
For example there was a suggestion above to name the struct members in test cases data structures. If we look at 
[this file](https://github.com/knative/serving/blob/77c03f0f6f5f86b0cfe5023786351974403276ec/pkg/queue/breaker_test.go)  or t[his one](https://github.com/knative/serving/blob/master/cmd/autoscaler/main_test.go) (I am sure there are more), this convention is not followed. @markusthoemmes thoughts?

markusthoemmes replied:
Yeah, pull them into a separate PR if possible, it only distracts from this one.

Personally, I usually do a rough pass at the thing I'm trying to solve and if I notice any refactorings I need, I create PRs for those that precede the actual future PR.

vagababov commented on pkg/autoscaler/metrics/http_scrape_client_test.go:
```diff
@@ -204,13 +237,61 @@ func TestHTTPScrapeClient_Scrape_ErrorCases(t *testing.T) {
 	}
 }
 
+func getProtoHTTPResponse(statusCode int, context Stat, contentType string) *http.Response {
```
```suggestion
func makeProtoResponse(statusCode int, stat Stat, contentType string) *http.Response {
```
Or `newProtoResponse`.
And I'd rename the other as well.

skonto replied:
I think then  it should be `makeHttpProtoResponse` because the other one cannot just be makeResponse.
It is good for readability to specify the type there.

vagababov replied:
it should bet `makeHTTP...` but sure.

vagababov commented on pkg/queue/breaker_test.go:
```diff
@@ -38,8 +38,8 @@ func TestBreakerInvalidConstructor(t *testing.T) {
 	}{{
-		"QueueDepth = 0",
-		BreakerParams{QueueDepth: 0, MaxConcurrency: 1, InitialCapacity: 1},
+		name:    "QueueDepth = 0",
```
if you went down this path, you need to finish it :)
But I think this does not belong to this pr.

skonto replied:
@vagababov I am not done yet :)

vagababov commented on pkg/queue/breaker_test.go:
```diff
@@ -86,7 +86,7 @@ func TestBreakerReserveOverload(t *testing.T) {
 
 func TestBreakerOverloadMixed(t *testing.T) {
-	// This tests when reservation and maybe are intermised.
+	// This testCases when reservation and maybe are intermised.
```
this change seems superfluous.

skonto replied:
Yeah I am not done yet this is refactoring relic.

vagababov commented on pkg/queue/breaker_test.go:
```diff
@@ -24,7 +24,7 @@ import (
 
 const (
-	// semAcquireTimeout is a timeout for tests that try to acquire
+	// semAcquireTimeout is a timeout for testCases that try to acquire
```
```suggestion
	// semAcquireTimeout is a timeout for tests that try to acquire
```

vagababov commented on pkg/queue/protobuf_stats_reporter.go:
```diff
@@ -0,0 +1,95 @@
+		if err != nil {
+			httpError(rsp, err)
+		}
+		header.Set(contentTypeHeader, "application/protobuf")
```
the const here.

vagababov commented on pkg/queue/protobuf_stats_reporter.go:
```diff
@@ -0,0 +1,95 @@
+	return http.HandlerFunc(func(rsp http.ResponseWriter, req *http.Request) {
+		stat := r.stat.Load()
+		if stat == nil {
+			httpError(rsp, errors.New("no metrics available yet."))
```
perhaps just pass the string since the only thing you do with the error is extract the string.
```suggestion
			httpError(rsp, errors.New("no metrics available yet"))
```

vagababov commented on pkg/queue/protobuf_stats_reporter_test.go:
```diff
@@ -0,0 +1,62 @@
+	}
+}
+
+func checkProtoData(t *testing.T, got float64, want float64) {
+	t.Helper()
+	if got != want {
+		t.Errorf("Got %v, wanted %v", got, want)
+	}
+}
```
1. This seems redundant.
2. It compares floats not protodata per se.

skonto commented on pkg/queue/protobuf_stats_reporter.go:
```diff
@@ -0,0 +1,95 @@
+)
+
+const (
+	contentEncodingHeader = "Content-Encoding"
```
Thanks I will remove this is wip anyway trying to do address the previous comments and improve coverage.

mattmoor commented on pkg/autoscaler/metrics/http_scrape_client.go:
```diff
@@ -20,6 +20,8 @@ import (
 	"errors"
 	"fmt"
 	"io"
+	"io/ioutil"
```
<!--gotool-bot-->
Format Go code:
```suggestion
	"io/ioutil"
	"net/http"

```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
	"io/ioutil"
	"net/http"

```

mattmoor commented on pkg/autoscaler/metrics/http_scrape_client.go:
```diff
@@ -20,6 +20,8 @@ import (
 	"io"
+	"io/ioutil"
+	"knative.dev/serving/pkg/network"
 	"net/http"
```
<!--gotool-bot-->
Format Go code:
```suggestion

```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion

```

mattmoor commented on pkg/autoscaler/metrics/http_scrape_client_test.go:
```diff
@@ -21,39 +21,67 @@ import (
 	"errors"
 	"fmt"
 	"io/ioutil"
+	"knative.dev/serving/pkg/network"
```
<!--gotool-bot-->
Format Go code:
```suggestion

```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion

```

mattmoor commented on pkg/autoscaler/metrics/http_scrape_client_test.go:
```diff
@@ -21,39 +21,67 @@ import (
+	"knative.dev/serving/pkg/network"
 	"net/http"
 	"strings"
 	"testing"
```
<!--gotool-bot-->
Format Go code:
```suggestion
	"testing"

	"knative.dev/serving/pkg/network"
```

mattmoor replied:
<!--gotool-bot-->
Format Go code:
```suggestion
	"testing"

	"knative.dev/serving/pkg/network"
```

skonto commented on pkg/queue/protobuf_stats_reporter_test.go:
```diff
@@ -0,0 +1,151 @@
+				}
+			}
+		})
+	}
```
Btw this part and any other part that iterates on structs could be compacted as follows but with very limited readability:
```
val1 := reflect.ValueOf(&stat).Elem()
val2 := reflect.ValueOf(&testStat).Elem()

for i := 0; i < val1.NumField(); i = i + 1 {
	val1.Field(i).Interface()
	if val1.Field(i).Interface() != val2.Field(i).Interface() {
		t.Errorf("handler returned wrong stat data: got %s = %v want %v", val1.Type().Field(i).Name,
		val1.Field(i).Interface(), val2.Field(i).Interface())
```

markusthoemmes replied:
Since I ditched `time.Time` off the metric, we can probably switch to using `cmp.Equal` and `cmp.Diff` here. They should do roughly what you've written out there.

skonto replied:
Yeah sure I guess they fail fast and report per field name.

julz commented on pkg/autoscaler/metrics/http_scrape_client.go:
```diff
@@ -53,11 +58,27 @@ func (c *httpScrapeClient) Scrape(url string) (Stat, error) {
-	return extractData(resp.Body)
+func statFromProto(body io.Reader) (Stat, error) {
+	var stat Stat
+	bodyBytes, err := ioutil.ReadAll(body)
```
this seems like it could cause some unfortunate GC load in large envs where there are lots of stats being gathered since we'll be creating a temporary buffer in each of these ReadAlls (and we already saw quite a lot of mem usage in stats scraping in our profiles). I don't know if it's worth a sync.Pool here, but it might be. Makes me think it'd be worth adding a benchmark for this path?

skonto replied:
Ok sure I will have a look.

skonto replied:
@julz @markusthoemmes  here are the [bench-marking results](https://gist.github.com/skonto/61c8343976e1840d44d89cf9f798f2aa) for using a pool. It uses less bytes obviously per op, but number of allocations is only one less.

markusthoemmes replied:
I think we can get away with a much smaller buffer here. In my preliminary tests the whole messages was just 66 bytes or so. Maybe we should try a few real world examples and see how big it actually gets. Maybe 128 or 256 bytes is far and enough here?

markusthoemmes replied:
But as I said, imo non-blocking for this PR.

julz replied:
(totally agree non-blocking for this PR)

skonto replied:
this is just an example i did try with 2048 at the beginning and since reducing it to 1024 didnt make much of a difference I just wanted to report the diff when using a pool. Yes a smaller buffer is the way to go.

skonto commented on cmd/queue/main.go:
```diff
@@ -420,15 +427,25 @@ func buildAdminServer(healthState *health.State, logger *zap.SugaredLogger) *htt
 }
 
+func metricsHttpHandler(promStatReporter *queue.PrometheusStatsReporter, protobufStatReporter *queue.ProtobufStatsReporter) http.Handler {
+	return http.HandlerFunc(func(rsp http.ResponseWriter, req *http.Request) {
```
Yeah ok.

skonto commented on cmd/queue/main.go:
```diff
@@ -420,15 +427,25 @@ func buildAdminServer(healthState *health.State, logger *zap.SugaredLogger) *htt
 
+func metricsHttpHandler(promStatReporter *queue.PrometheusStatsReporter, protobufStatReporter *queue.ProtobufStatsReporter) http.Handler {
+	return http.HandlerFunc(func(rsp http.ResponseWriter, req *http.Request) {
+		if strings.Contains(req.Header.Get("Accept"), network.ProtoAcceptContent) {
```
Potentially this could be a list of accepted contents, so it is more technically correct to do it this way. However, in this context it should be one type of content in general, because prometheus text format didnt require to set anything.

markusthoemmes commented on cmd/queue/main.go:
```diff
@@ -232,13 +233,19 @@ func main() {
 	}
 
+	protoStatReporter := queue.NewProtobufStatsReporter(env.ServingPod, reportingPeriod)
+	if err != nil {
+		logger.Fatalw("Failed to create stats reporter", zap.Error(err))
+	}
```
Those are no longer necessary.

markusthoemmes commented on pkg/queue/protobuf_stats_reporter.go:
```diff
@@ -0,0 +1,92 @@
+	return http.HandlerFunc(func(rsp http.ResponseWriter, req *http.Request) {
+		stat := r.stat.Load()
+		if stat == nil {
+			httpError(rsp, "no metrics available yet")
```
How does prometheus behave here? Should we return an empty metric instead?

skonto replied:
I had a look at this before returning this error, so the code that does check metrics returned is [here](https://github.com/prometheus/client_golang/blob/d3accfda2168b0e253f98cff3ef81586e8b88d9f/prometheus/promhttp/http.go#L126-L142). If things fail it returns an http error.

skonto commented:
@vagababov @markusthoemmes @taragu addressed all comments.
Things I see pending before LGTM:
a) Fix struct comparison stuff.
b) Add a pool for reading http body.
c) Investigate why `TestTargetBurstCapacity` fails at first run, (seems irrelevant to me but I also have limited knowledge so maybe something delays the test, not sure or something else is lurking here that emerged due to this PR).

Regarding the coverage diff this is because as previously error cases are not tested (as with prometheus path it does not make much difference to test them), report attached here with what not covered: 
![cov1](https://user-images.githubusercontent.com/7945591/86805941-0c6cea00-c081-11ea-9411-46130e52eb4b.png)

markusthoemmes commented:
The coverage delta seems fine and the pooling can easily be a TODO for now imo. We'll have the entire release cycle to improve on it. The same with struct comparison, if you don't want to fix it in this PR.

The most critical is stable tests :).

vagababov commented:
This seems to have broken Target burst capacity integration test
Somehow...

skonto commented:
@vagababov @markusthoemmes before my lqst update that test didn't fail the second time. Failure based on a timeout seems unrelated at a first glance, usually its because something is very slow maybe in the tets env. I will debug locally but i was hoping for some insights from your side.

skonto commented:
/retest

vagababov commented:
Definitely something's fishy.

markusthoemmes commented on cmd/queue/main.go:
```diff
@@ -232,13 +233,16 @@ func main() {
 	stats := network.NewRequestStats(time.Now())
 	go func() {
 		for now := range reportTicker.C {
 			promStatReporter.Report(stats.Report(now))
+			protoStatReporter.Report(stats.Report(now))
```
`stats.Report` resets the stat reporting. You'll want:

```golang
stat := stats.Report(now)
promStatReporter.Report(stat)
protoStatReporter.Report(stat)
```

I'm fairly sure that's the bug that causes the test failure.

skonto replied:
Yeah I was thinking of that part because if it does not get the right stats it will not scale. Let me try it.

markusthoemmes commented on pkg/queue/protobuf_stats_reporter_test.go:
```diff
@@ -0,0 +1,151 @@
+			if err != nil {
+				t.Fatal(err)
+			}
+			req.Header.Set("Accept-content", network.ProtoAcceptContent)
```
What's this for? I don't think we need the header at all and `Accept-content` seems used nowhere else.

skonto replied:
I am trying to mimic the real http request which contains the header, it can be omitted since the tests just exercises the proto handler and the handler assumes that field has been tested earlier.

markusthoemmes replied:
Sure we can keep it, but please make it `Accept` then as all other parts of the system do too, I believe.

markusthoemmes commented on pkg/queue/protobuf_stats_reporter_test.go:
```diff
@@ -0,0 +1,151 @@
+			startTime:       time.Now(),
+			stat:            metricsStat,
+			podName:         "testPod"},
+	},
+	}
```
nit: `}}`

markusthoemmes commented on pkg/queue/protobuf_stats_reporter_test.go:
```diff
@@ -0,0 +1,151 @@
+			} else { // good case, data received
+				bodyBytes, err := ioutil.ReadAll(rr.Body)
+				if err != nil {
+					t.Errorf("reading body failed: %w", err)
```
I think the `%w` will do odd things here. It should be `%v`. `%w` only works for `fmt.Errorf`

markusthoemmes commented on pkg/queue/protobuf_stats_reporter_test.go:
```diff
@@ -0,0 +1,151 @@
+				stat := metrics.Stat{}
+				err = stat.Unmarshal(bodyBytes)
+				if err != nil {
+					t.Errorf("unmarshalling failed: %w", err)
```
Same as above.

markusthoemmes commented on pkg/autoscaler/metrics/http_scrape_client.go:
```diff
@@ -65,9 +76,20 @@ func (c *httpScrapeClient) Scrape(url string) (Stat, error) {
+func ioCopy(r io.Reader) ([]byte, error) {
+	buffer := pool.Get().(*bytes.Buffer)
+	buffer.Reset()
+	defer pool.Put(buffer)
```
I don't think this is safe. We can only return the buffer after we've unmarshalled the stat. The underlying byte array will still be needed after this function.

I'd recommend to back out the bufferpool stuff from this PR honestly, it unnecessarily complicates things.

skonto replied:
Ok I will remove but it could be fixed easily by not calling the function,and move its contents in the statFromProto fucntion.

markusthoemmes replied:
sure but why not land the additional improvement in a followup PR?

julz replied:
Yeah, also [it's not really safe](https://github.com/golang/go/issues/23199) to store a bytes.Buffer in a sync.Pool. 

Agree with @markusthoemmes this is overkill in this PR anyway, let's look at it as a follow-up: I mostly just wanted a Benchmark so we were sure this actually ends up being an improvement with the extra allocations.

skonto replied:
a@julz we could return an array of fixed size if I read correctly to solve this, in addition the writes here are small. Anyway I removed that part.

markusthoemmes commented on pkg/queue/protobuf_stats_reporter_test.go:
```diff
@@ -24,18 +24,28 @@ import (
+		ProcessUptime:                    20.0,
+	}
+
+	ignoreStuff = cmp.Options{
+		cmpopts.IgnoreFields(testStat, "ProcessUptime"),
+	}
```
Why do we ignore this? The old test code had a hardcoded `==` check here :thinking:

skonto replied:
Misread about the time discussion above.

markusthoemmes commented on pkg/queue/protobuf_stats_reporter.go:
```diff
@@ -0,0 +1,92 @@
+		data := stat.(metrics.Stat)
+		buffer, err := proto.Marshal(&data)
+		if err != nil {
+			httpError(rsp, err.Error())
```
Sorry for the late comment: Do we need a `return` after this?

skonto replied:
Yes just fixed it.

skonto commented:
@markusthoemmes thanks. @vagababov I think this is ready now. WDYT?

vagababov commented:
/lgtm

vagababov commented:
/unhold
